app/routes.py

```diff
@@ -1,2 +1,219 @@
+    else:
+        tasks = Task.query.all()
+    
+    filter_query = request.args.get("title")
+    if filter_query:
+        tasks = Task.query.filter(Task.title.contains(filter_query))
+    
+    return tasks
+
+@tasks_bp.route("", methods=["GET"])
+def get_tasks():
+    tasks = sort_filter_tasks()
+
+    tasks_response = [task.to_dict() for task in tasks]
+
+    return jsonify(tasks_response), 200
+
+def sort_filter_goals():
+    sort_query = request.args.get("sort")
+    if sort_query == "desc":
+        goals = Goal.query.order_by(Goal.title.desc())
+    elif sort_query == "asc":
+        goals = Goal.query.order_by(Goal.title.asc())
+    else:
+        goals = Goal.query.all()
+
+    filter_query = request.args.get("title")
+    if filter_query:
+        goals = Goal.query.filter(Goal.title.contains(filter_query))
+    
+    return goals
+
+@goals_bp.route("", methods=["GET"])
+def get_goals():
+    goals = sort_filter_goals()
+
+    goal_response = [goal.to_dict() for goal in goals]
+
+    return jsonify(goal_response), 200
+
+@goals_bp.route("/<goal_id>", methods=["GET"])
+@require_goal
+def get(goal):  
+    return jsonify({"goal": goal.to_dict()}), 200
+
+@tasks_bp.route("/<task_id>", methods=["GET"])
+@require_task
+def get(task):  
+    if task.goal_id:
+        return jsonify({"task": task.task_to_dict_w_goal()}), 200
+    return jsonify({"task": task.to_dict()}), 200
+
+@goals_bp.route("/<goal_id>/tasks", methods=["GET"])
+@require_goal
+def get_tasked_goal(goal):
+    return jsonify(goal.goal_task_dict()), 200
+
+@goals_bp.route("/<goal_id>", methods=["PUT"])
+@require_goal
+def put(goal):
+    form_data = request.get_json()
+
+    goal.replace_with_dict(form_data)
+
+    db.session.commit()
+
+    return jsonify({"goal": goal.to_dict()}), 200
+
+@tasks_bp.route("/<task_id>", methods=["PUT"])
+@require_task
+def put(task):
+    form_data = request.get_json()
+
+    task.replace_with_dict(form_data)
+
+    db.session.commit()
+
+    return jsonify({"task": task.to_dict()}), 200
+
+@goals_bp.route("/<goal_id>", methods=["DELETE"])
+@require_goal
+def delete(goal):
+    db.session.delete(goal)
+    db.session.commit()
+    return jsonify ({"details": (f"Goal {goal.id} \"{goal.title}\" successfully deleted")}), 200
+
+@tasks_bp.route("/<task_id>", methods=["DELETE"])
+@require_task
+def delete(task):
+    db.session.delete(task)
+    db.session.commit()
+    return jsonify ({"details": (f"Task {task.id} \"{task.title}\" successfully deleted")}), 200
+
+@tasks_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+@require_task
+def complete_patch(task):
+    request.get_json()
+
+    task.completed_at = datetime.utcnow()
+
+    db.session.commit()
+
+    load_dotenv()
+
+    data = {"token": os.environ.get("SLACK_TOKEN"), 
+                "channel": os.environ.get("CHANNEL_ID"), 
+                "text": f"Someone just completed the task {task.title}"}
+    url = os.environ.get("SLACK_URL")
+    requests.post(url, data)
```

Comment:
Consider putting this part of the code into its own helper function. All the mark complete endpoint really needs to know how to do is call a method that can post to Slack with the message to post. The rest of the details could be hidden in the helper, and potentially called from multiple places!